### PR TITLE
requires/types: introduce expression type (expr.py)

### PR DESCRIPTION
**Pull request by xmkg**
`expression` is a new requirement type that allows users to express a requirement in human-readable, SQL-like syntax. It has proper grammar to allow users to write checks in as easily digestible manner, which will improve the QOL for scenario writers and code reviewers. The grammar is constructed using the `pyparsing` library.

Example:
```yaml
checks:
  not_using_iommu_passthrough_for_suitable_amd: |
    (@hotsos.core.plugins.kernel.sysfs.CPU.vendor == 'authenticamd') AND
    (@hotsos.core.plugins.system.SystemBase.virtualisation_type == None) AND
    NOT('iommu=pt' IN @hotsos.core.plugins.kernel.KernelBase.boot_parameters)    
```

With the [alias patch](https://github.com/canonical/hotsos/pull/929), it'll look much prettier:

```yaml
checks:
  not_using_iommu_passthrough_for_suitable_amd: |
    (@kernel.sysfs.cpu.vendor == 'authenticamd') AND
    (@system.virtualisation_type == None) AND
    NOT('iommu=pt' IN @kernel.boot_parameters)    
```

The current grammar supports the following constructs:

- Keywords (True, False, None)
- Constants (Integer, Float, String literal)
- Runtime variables (Python properties)
- Functions(len, not, file, systemd, read_ini, read_cert)
- Arithmetic operators(sign, mul, div, add, sub, exp)
- Comparison operators(<, <=, >, >=, ==, !=, in)
- Logical operators(and, or, not)
- Comments('#', '//', '/*...*/')

Also, the following changes have been made:

- Updated requirement_types.rst to document the new Expression requirement type.
- Moved property resolver logic from YPropertyBase to PythonEntityResolver class.
- Added `pyparsing` as a dependency.
- Added unit tests for the new code.

This patch rewrites some of the scenario checks in the new expression syntax to demonstrate the difference and provide examples.

### CODE REVIEW

**Comment by mustafakemalgilor** [comment]
IDK why the "Test / test (3.8, ubuntu-20.04) (pull_request)" job is failing, all tests are green in my env. I'll re-run them in a py3.8 environment and get to the bottom of it.

**Comment by mustafakemalgilor** [comment]
@dosaboy fixed CI issues, ready for review. please feel free to ping me should you need to ask anything.

**Comment by mustafakemalgilor** [comment]
This feature is intended to be used in conjunction with the aliasing feature: https://github.com/canonical/hotsos/pull/929

**Review by brianphaley — APPROVED** [review]

File: doc/source/contrib/language_ref/property_ref/requirement_types.rst [hunk 1]
```diff
@@ -337,3 +337,622 @@ Cache keys:
 * input_value: value of the variable used as input
 * ops: str representation of ops list
 
+
+Expression
+----------
+
+Expressions allows the user to express a requirement in a human-readable domain specific language. It's purpose
```
Comment (nicolasbock) [inline doc/source/contrib/language_ref/property_ref/requirement_types.rst:344]:
```suggestion
Expressions allow the user to express a requirement in a human-readable domain specific language. Its purpose
```

File: doc/source/contrib/language_ref/property_ref/requirement_types.rst [hunk 2]
```diff
@@ -337,3 +337,622 @@ Cache keys:
...
+is to provide a simplified, clear an concise expression grammar to all requirement types, and eliminate redundancies.
```
Comment (nicolasbock) [inline doc/source/contrib/language_ref/property_ref/requirement_types.rst:345]:
```suggestion
is to provide a simplified, clear and concise expression grammar for all requirement types and to eliminate redundancies.
```

File: hotsos/core/plugins/kernel/memory.py [hunk 3]
```diff
@@ -106,11 +106,13 @@ def huge_pages_enabled(self):
 
     @property
     def hugetlb_to_mem_total_percentage(self):
-        return round((self.Hugetlb * 100) / self.MemTotal)
+        return (self.MemTotal and round(
```
Comment (nicolasbock) [inline hotsos/core/plugins/kernel/memory.py:109]: Why add `or 0`?

**Comment by xmkg** [comment]
@dosaboy what's your opinion on this? Let me know should you decide to land this and I'll make the all required changes, should be there any.

**Pull request closed without merging**